File: naddu/naddu_2/c/history.c
#include <string.h>
#include "history.h"

int16_t piece_history[16][128];

void history_init(void) {
  history_clear();
}

void history_clear(void) {
  memset(piece_history, 0, sizeof(piece_history));
}

void history_halve(void) {
  for (int i = 0; i < 16; i++) {
    for (int j = 0; j < 128; j++) {
      piece_history[i][j] >>= 1;
    }
  }
}

void add_history(const Pos *pos, uint32_t move, int depth) {
  if (move & MOVE_FLAG_CAPTURE)
    return;

  int from = MOVE_FROM(move);
  int to = MOVE_TO(move);
  int piece = pos->board[from];
  int bonus = depth * depth;
  int current = piece_history[piece][to];
  int update = bonus - (current * bonus) / HISTORY_MAX;
  int new_val = current + update;

  if (new_val >= HISTORY_MAX) {
    history_halve();
    piece_history[piece][to] += update / 2;
  }
  else {
    piece_history[piece][to] = new_val;
  }
}

File: naddu/naddu_2/src/history.js

const pieceHistory = [];
const HISTORY_MAX = 32767 - MAX_PLY * MAX_PLY;

function historyInitOnce() {
  for (let i = 0; i < 16; i++) {
    pieceHistory[i] = new Int16Array(128);
  }
}

function historyClear() {
  for (let i = 0; i < 16; i++) {
    pieceHistory[i].fill(0);
  }
}

function historyHalve() {
  for (let i = 0; i < 16; i++) {
    for (let j = 0; j < 128; j++) {
      pieceHistory[i][j] = pieceHistory[i][j] >> 1;
    }
  }
}

function addHistory(pos, move, depth) {
  if (move & MOVE_FLAG_CAPTURE)
    return;
  const from = (move >> 8) & 0xff;
  const to = move & 0xff;
  const piece = pos.board[from];
  const bonus = depth * depth;
  const current = pieceHistory[piece][to];
  const update = bonus - ((current * bonus) / HISTORY_MAX) | 0;
  const newVal = current + update;
  if (newVal >= HISTORY_MAX) {
    historyHalve();
    pieceHistory[piece][to] += (update / 2) | 0;
  }
  else {
    pieceHistory[piece][to] = newVal;
  }
}
